File: services/observability-service/src/middleware/security.ts
import { Request, Response, NextFunction } from 'express';

// Security event types for observability service
enum ObservabilitySecurityEventType {
  METRICS_ACCESS = 'metrics_access',
  BILLING_ACCESS = 'billing_access',
  USAGE_EVENT_INGESTION = 'usage_event_ingestion',
  UNAUTHORIZED_ACCESS = 'unauthorized_access',
  RATE_LIMIT_EXCEEDED = 'rate_limit_exceeded',
  SUSPICIOUS_ACTIVITY = 'suspicious_activity',
  DATA_VALIDATION_FAILED = 'data_validation_failed',
  XSS_ATTEMPT = 'xss_attempt',
  SQL_INJECTION_ATTEMPT = 'sql_injection_attempt',
  METRICS_SCRAPING = 'metrics_scraping',
  BILLING_MANIPULATION_ATTEMPT = 'billing_manipulation_attempt',
  EXCESSIVE_EVENTS = 'excessive_events'
}

// Security context interface
interface ObservabilitySecurityContext {
  userId: string;
  action: string;
  ip: string;
  userAgent: string;
  timestamp: Date;
  riskScore: number;
  eventType: ObservabilitySecurityEventType;
  metadata?: Record<string, any>;
}

// In-memory storage for rate limiting and security tracking
class ObservabilitySecurityStore {
  private rateLimitStore = new Map<string, { count: number; resetTime: number }>();
  private securityEvents: ObservabilitySecurityContext[] = [];
  private suspiciousIPs = new Set<string>();
  private blockedIPs = new Set<string>();
  private eventCounts = new Map<string, { count: number; window: number }>();
  
  // Track rate limits with different limits for different endpoints
  trackRequest(key: string, windowMs: number, maxRequests: number): { allowed: boolean; remaining: number } {
    const now = Date.now();
    const record = this.rateLimitStore.get(key);
    
    if (!record || now > record.resetTime) {
      this.rateLimitStore.set(key, { count: 1, resetTime: now + windowMs });
      return { allowed: true, remaining: maxRequests - 1 };
    }
    
    record.count++;
    this.rateLimitStore.set(key, record);
    return { allowed: record.count <= maxRequests, remaining: Math.max(0, maxRequests - record.count) };
  }
  
  // Track event ingestion for monitoring abuse
  trackEventIngestion(ip: string): { allowed: boolean; count: number } {
    const now = Date.now();
    const windowMs = 60 * 1000; // 1 minute window
    const maxEvents = 100; // Max 100 events per minute per IP
    
    const key = `events:${ip}`;
    const record = this.eventCounts.get(key);
    
    if (!record || now > record.window) {
      this.eventCounts.set(key, { count: 1, window: now + windowMs });
      return { allowed: true, count: 1 };
    }
    
    record.count++;
    this.eventCounts.set(key, record);
    return { allowed: record.count <= maxEvents, count: record.count };
  }
  
  // Log security events
  logSecurityEvent(event: ObservabilitySecurityContext): void {
    this.securityEvents.push(event);
    
    // Keep only last 1000 events to prevent memory issues
    if (this.securityEvents.length > 1000) {
      this.securityEvents.shift();
    }
    
    // Auto-block IPs with too many suspicious events
    const recentSuspiciousEvents = this.securityEvents.filter(
      e => e.ip === event.ip && 
      e.timestamp.getTime() > Date.now() - 300000 && // Last 5 minutes
      e.riskScore > 7
    );
    
    if (recentSuspiciousEvents.length >= 3) {
      this.blockedIPs.add(event.ip);
      console.warn(`Auto-blocked IP ${event.ip} due to suspicious observability activity`);
    }
  }
  
  // Get security metrics
  getMetrics() {
    const now = Date.now();
    const last24h = this.securityEvents.filter(e => e.timestamp.getTime() > now - 86400000);
    
    return {
      totalEvents: this.securityEvents.length,
      eventsLast24h: last24h.length,
      blockedIPs: Array.from(this.blockedIPs),
      suspiciousIPs: Array.from(this.suspiciousIPs),
      eventsByType: this.groupEventsByType(last24h),
      highRiskEvents: last24h.filter(e => e.riskScore > 8).length,
      activeRateLimits: this.rateLimitStore.size,
      timestamp: new Date().toISOString()
    };
  }
  
  private groupEventsByType(events: ObservabilitySecurityContext[]) {
    return events.reduce((acc, event) => {
      acc[event.eventType] = (acc[event.eventType] || 0) + 1;
      return acc;
    }, {} as Record<string, number>);
  }
  
  // Check if IP is blocked
  isBlocked(ip: string): boolean {
    return this.blockedIPs.has(ip);
  }
  
  // Mark IP as suspicious
  markSuspicious(ip: string): void {
    this.suspiciousIPs.add(ip);
  }
}

// Main security middleware class for observability service
export class ObservabilitySecurityMiddleware {
  private store = new ObservabilitySecurityStore();
  
  // Security headers middleware
  observabilitySecurityHeaders() {
    return (_req: Request, res: Response, next: NextFunction) => {
      // Enhanced security headers for observability service
      res.setHeader('X-Content-Type-Options', 'nosniff');
      res.setHeader('X-Frame-Options', 'DENY');
      res.setHeader('X-XSS-Protection', '1; mode=block');
      res.setHeader('Strict-Transport-Security', 'max-age=31536000; includeSubDomains');
      res.setHeader('Referrer-Policy', 'strict-origin-when-cross-origin');
      res.setHeader('Content-Security-Policy', 
        "default-src 'self'; " +
        "script-src 'self'; " +
        "style-src 'self'; " +
        "img-src 'self'; " +
        "connect-src 'self'; " +
        "font-src 'self'; " +
        "object-src 'none'; " +
        "media-src 'self'; " +
        "frame-src 'none';"
      );
      res.setHeader('Permissions-Policy', 
        'geolocation=(), microphone=(), camera=(), payment=(), usb=(), magnetometer=(), gyroscope=()'
      );
      
      next();
    };
  }
  
  // Rate limiting middleware with endpoint-specific limits
  observabilityRateLimit() {
    return (req: Request, res: Response, next: NextFunction) => {
      const ip = this.getClientIP(req);
      
      // Check if IP is blocked
      if (this.store.isBlocked(ip)) {
        this.logSecurityEvent(req, ObservabilitySecurityEventType.RATE_LIMIT_EXCEEDED, 10, {
          reason: 'IP blocked due to previous violations',
          blockedIP: ip
        });
        return res.status(429).json({ 
          error: 'Too many requests - IP temporarily blocked',
          retryAfter: '1 hour'
        });
      }
      
      // Different rate limits for different endpoints
      let maxRequests = 100;
      let windowMinutes = 15;
      
      if (req.path.includes('/metrics')) {
        maxRequests = 200; // Prometheus scraping needs higher limits
        windowMinutes = 5;
      } else if (req.path.includes('/billing')) {
        maxRequests = 30; // Billing operations should be limited
        windowMinutes = 15;
      } else if (req.path.includes('/usage/event')) {
        // Use special handling for event ingestion
        const eventCheck = this.store.trackEventIngestion(ip);
        if (!eventCheck.allowed) {
          this.logSecurityEvent(req, ObservabilitySecurityEventType.EXCESSIVE_EVENTS, 8, {
            eventCount: eventCheck.count,
            limit: 100,
            window: '1 minute'
          });
          
          return res.status(429).json({ 
            error: 'Event ingestion rate limit exceeded',
            limit: 100,
            window: '1 minute'
          });
        }
        return next(); // Skip regular rate limiting for event ingestion
      }
      
      const key = `rate_limit:${ip}:${req.path}`;
      const windowMs = windowMinutes * 60 * 1000;
      const result = this.store.trackRequest(key, windowMs, maxRequests);
      
      // Add rate limit headers
      res.setHeader('X-RateLimit-Limit', maxRequests.toString());
      res.setHeader('X-RateLimit-Remaining', result.remaining.toString());
      res.setHeader('X-RateLimit-Reset', new Date(Date.now() + windowMs).toISOString());
      
      if (!result.allowed) {
        this.logSecurityEvent(req, ObservabilitySecurityEventType.RATE_LIMIT_EXCEEDED, 6, {
          limit: maxRequests,
          window: windowMinutes,
          endpoint: req.path
        });
        
        return res.status(429).json({ 
          error: 'Rate limit exceeded',
          limit: maxRequests,
          window: `${windowMinutes} minutes`,
          retryAfter: Math.ceil(windowMs / 1000)
        });
      }
      
      next();
    };
  }
  
  // Data validation middleware
  validateObservabilityData() {
    return (req: Request, res: Response, next: NextFunction) => {
      try {
        // Input sanitization and validation
        if (req.body) {
          const sanitized = this.sanitizeInput(req.body);
          req.body = sanitized;
          
          // Check for malicious patterns
          const maliciousContent = this.detectMaliciousContent(JSON.stringify(req.body));
          if (maliciousContent.detected) {
            this.logSecurityEvent(req, maliciousContent.type, 9, {
              maliciousPattern: maliciousContent.pattern,
              sanitizedData: 'blocked'
            });
            
            return res.status(400).json({ 
              error: 'Invalid input detected',
              code: 'VALIDATION_FAILED'
            });
          }
          
          // Validate observability-specific data
          const validationResult = this.validateObservabilityPayload(req.body, req.path);
          if (!validationResult.valid) {
            this.logSecurityEvent(req, ObservabilitySecurityEventType.DATA_VALIDATION_FAILED, 5, {
              errors: validationResult.errors,
              endpoint: req.path
            });
            
            return res.status(400).json({
              error: 'Data validation failed',
              details: validationResult.errors
            });
          }
        }
        
        next();
      } catch (error) {
        console.error('Observability data validation error:', error);
        this.logSecurityEvent(req, ObservabilitySecurityEventType.DATA_VALIDATION_FAILED, 7, {
          error: error instanceof Error ? error.message : 'Unknown validation error'
        });
        
        res.status(500).json({ error: 'Validation error occurred' });
      }
    };
  }
  
  // Audit logging for observability operations
  auditObservabilityOperations() {
    const self = this;
    return (req: Request, res: Response, next: NextFunction) => {
      const originalSend = res.json;
      
      res.json = function(body: any) {
        // Determine event type based on request
        let eventType = ObservabilitySecurityEventType.METRICS_ACCESS;
        let riskScore = 1;
        
        if (req.path.includes('/metrics')) {
          eventType = ObservabilitySecurityEventType.METRICS_SCRAPING;
          riskScore = 1; // Normal operation
        } else if (req.path.includes('/billing')) {
          eventType = ObservabilitySecurityEventType.BILLING_ACCESS;
          riskScore = 3; // Higher risk due to financial data
          
          // Check for suspicious billing patterns
          if (req.body?.amount && req.body.amount > 10000) {
            eventType = ObservabilitySecurityEventType.BILLING_MANIPULATION_ATTEMPT;
            riskScore = 9;
          }
        } else if (req.path.includes('/usage/event')) {
          eventType = ObservabilitySecurityEventType.USAGE_EVENT_INGESTION;
          riskScore = 2;
        }
        
        // Log the operation
        self.logSecurityEvent(req, eventType, riskScore, {
          action: `${req.method} ${req.path}`,
          statusCode: res.statusCode,
          responseSize: JSON.stringify(body).length,
          requestBody: req.method !== 'GET' ? req.body : undefined
        });
        
        return originalSend.call(this, body);
      };
      
      next();
    };
  }
  
  // Metrics endpoint protection
  protectMetricsEndpoint() {
    return (req: Request, res: Response, next: NextFunction) => {
      const ip = this.getClientIP(req);
      
      // Allow Prometheus-like user agents and internal requests
      const userAgent = req.headers['user-agent'] || '';
      const isPrometheus = userAgent.includes('Prometheus') || userAgent.includes('prometheus');
      const isInternal = ip.startsWith('127.') || ip.startsWith('10.') || ip.startsWith('192.168.');
      
      if (!isPrometheus && !isInternal) {
        // Log external metrics access
        this.logSecurityEvent(req, ObservabilitySecurityEventType.METRICS_ACCESS, 4, {
          userAgent,
          ip,
          reason: 'External metrics access'
        });
        
        // Add warning header but allow access
        res.setHeader('X-Metrics-Warning', 'External access to metrics endpoint');
      }
      
      next();
    };
  }
  
  // Get security metrics
  async getSecurityMetrics() {
    return this.store.getMetrics();
  }
  
  // Helper methods
  private getClientIP(req: Request): string {
    return (req.headers['x-forwarded-for'] as string)?.split(',')[0] ||
           req.headers['x-real-ip'] as string ||
           req.socket.remoteAddress ||
           'unknown';
  }
  
  private sanitizeInput(input: any): any {
    if (typeof input === 'string') {
      // Basic XSS protection
      return input
        .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
        .replace(/<iframe\b[^<]*(?:(?!<\/iframe>)<[^<]*)*<\/iframe>/gi, '')
        .replace(/javascript:/gi, '')
        .replace(/on\w+=/gi, '')
        .trim();
    }
    
    if (Array.isArray(input)) {
      return input.map(item => this.sanitizeInput(item));
    }
    
    if (input && typeof input === 'object') {
      const sanitized: any = {};
      for (const [key, value] of Object.entries(input)) {
        sanitized[key] = this.sanitizeInput(value);
      }
      return sanitized;
    }
    
    return input;
  }
  
  private detectMaliciousContent(content: string): { detected: boolean; type: ObservabilitySecurityEventType; pattern?: string } {
    const patterns = [
      { pattern: /<script|javascript:|on\w+=/i, type: ObservabilitySecurityEventType.XSS_ATTEMPT },
      { pattern: /(\b(union|select|insert|delete|update|drop|create|alter|exec|execute)\b)|(-{2})|(\|\|)/i, type: ObservabilitySecurityEventType.SQL_INJECTION_ATTEMPT },
      { pattern: /\.\.(\/|\\)/g, type: ObservabilitySecurityEventType.SUSPICIOUS_ACTIVITY }
    ];
    
    for (const { pattern, type } of patterns) {
      if (pattern.test(content)) {
        return { detected: true, type, pattern: pattern.toString() };
      }
    }
    
    return { detected: false, type: ObservabilitySecurityEventType.METRICS_ACCESS };
  }
  
  private validateObservabilityPayload(data: any, endpoint: string): { valid: boolean; errors: string[] } {
    const errors: string[] = [];
    
    if (endpoint.includes('/usage/event')) {
      // Validate usage event data
      if (!data.eventType || typeof data.eventType !== 'string') {
        errors.push('eventType is required and must be a string');
      }
      
      if (!data.userId || typeof data.userId !== 'string') {
        errors.push('userId is required and must be a string');
      }
      
      if (data.timestamp && !(data.timestamp instanceof Date) && isNaN(Date.parse(data.timestamp))) {
        errors.push('timestamp must be a valid date');
      }
      
      if (data.metadata && typeof data.metadata !== 'object') {
        errors.push('metadata must be an object');
      }
    }
    
    if (endpoint.includes('/billing')) {
      // Validate billing data
      if (data.amount !== undefined) {
        if (typeof data.amount !== 'number' || data.amount < 0 || data.amount > 100000) {
          errors.push('amount must be a number between 0 and 100000');
        }
      }
      
      if (data.currency && typeof data.currency !== 'string') {
        errors.push('currency must be a string');
      }
      
      if (data.userId && typeof data.userId !== 'string') {
        errors.push('userId must be a string');
      }
    }
    
    return { valid: errors.length === 0, errors };
  }
  
  private logSecurityEvent(
    req: Request, 
    eventType: ObservabilitySecurityEventType, 
    riskScore: number, 
    metadata: Record<string, any> = {}
  ): void {
    const context: ObservabilitySecurityContext = {
      userId: 'system', // Observability service typically doesn't have user context
      action: `${req.method} ${req.path}`,
      ip: this.getClientIP(req),
      userAgent: req.headers['user-agent'] || '',
      timestamp: new Date(),
      riskScore,
      eventType,
      metadata: {
        ...metadata,
        query: req.query,
        headers: {
          'content-type': req.headers['content-type'],
          'authorization': req.headers.authorization ? '[REDACTED]' : undefined
        }
      }
    };
    
    this.store.logSecurityEvent(context);
    
    // Log high-risk events to console
    if (riskScore > 7) {
      console.warn(`High-risk observability security event: ${eventType}`, {
        ip: context.ip,
        riskScore,
        metadata
      });
    }
  }
}

export const observabilitySecurityMiddleware = new ObservabilitySecurityMiddleware();
export { ObservabilitySecurityEventType, ObservabilitySecurityContext };

File: services/grammar-service/app/security.py
"""
Grammar Service Security Middleware
Provides comprehensive security features for the grammar evaluation service
"""

import asyncio
import time
import re
import logging
import json
from typing import Dict, Any, Optional, Set, List
from collections import defaultdict, deque
from datetime import datetime, timedelta
from enum import Enum
from dataclasses import dataclass, asdict
from fastapi import Request, Response, HTTPException
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
import hashlib
import html

# Configure logging
logger = logging.getLogger("grammar_security")
logger.setLevel(logging.INFO)

class GrammarSecurityEventType(Enum):
    """Security event types for grammar service"""
    GRAMMAR_EVALUATION = "grammar_evaluation"
    RATE_LIMIT_EXCEEDED = "rate_limit_exceeded"
    UNAUTHORIZED_ACCESS = "unauthorized_access" 
    SUSPICIOUS_ACTIVITY = "suspicious_activity"
    DATA_VALIDATION_FAILED = "data_validation_failed"
    XSS_ATTEMPT = "xss_attempt"
    CONTENT_FILTERING = "content_filtering"
    MALICIOUS_CONTENT = "malicious_content"
    TEXT_ABUSE = "text_abuse"
    LARGE_PAYLOAD = "large_payload"
    LANGUAGE_MANIPULATION = "language_manipulation"

class RiskLevel(Enum):
    """Risk levels for security events"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

@dataclass
class SecurityContext:
    """Security context for tracking requests"""
    ip: str
    user_agent: str
    timestamp: datetime
    endpoint: str
    method: str
    risk_score: int
    event_type: GrammarSecurityEventType
    metadata: Dict[str, Any]

class InMemorySecurityStore:
    """In-memory storage for security tracking"""
    
    def __init__(self):
        self.rate_limits: Dict[str, Dict[str, Any]] = {}
        self.security_events: deque = deque(maxlen=1000)  # Keep last 1000 events
        self.blocked_ips: Set[str] = set()
        self.suspicious_ips: Set[str] = set()
        self.content_hashes: Set[str] = set()  # Track repeated content
        self.user_requests: defaultdict = defaultdict(list)  # Track user request patterns
        
    def track_request(self, key: str, window_seconds: int, max_requests: int) -> Dict[str, Any]:
        """Track rate limiting"""
        now = time.time()
        
        if key not in self.rate_limits:
            self.rate_limits[key] = {"count": 1, "reset_time": now + window_seconds}
            return {"allowed": True, "remaining": max_requests - 1}
            
        limit_data = self.rate_limits[key]
        if now > limit_data["reset_time"]:
            # Reset window
            self.rate_limits[key] = {"count": 1, "reset_time": now + window_seconds}
            return {"allowed": True, "remaining": max_requests - 1}
            
        limit_data["count"] += 1
        allowed = limit_data["count"] <= max_requests
        remaining = max(0, max_requests - limit_data["count"])
        
        return {"allowed": allowed, "remaining": remaining}
        
    def log_security_event(self, context: SecurityContext):
        """Log security event"""
        self.security_events.append(context)
        
        # Auto-block IPs with high-risk activities
        if context.risk_score >= 8:
            recent_events = [
                e for e in self.security_events 
                if e.ip == context.ip and 
                (datetime.now() - e.timestamp).seconds < 300  # Last 5 minutes
                and e.risk_score >= 7
            ]
            
            if len(recent_events) >= 3:
                self.blocked_ips.add(context.ip)
                logger.warning(f"Auto-blocked IP {context.ip} due to suspicious activity")
                
    def is_blocked(self, ip: str) -> bool:
        """Check if IP is blocked"""
        return ip in self.blocked_ips
        
    def get_metrics(self) -> Dict[str, Any]:
        """Get security metrics"""
        now = datetime.now()
        last_24h = [
            e for e in self.security_events 
            if (now - e.timestamp).total_seconds() < 86400
        ]
        
        return {
            "total_events": len(self.security_events),
            "events_last_24h": len(last_24h),
            "blocked_ips": list(self.blocked_ips),
            "suspicious_ips": list(self.suspicious_ips),
            "high_risk_events": len([e for e in last_24h if e.risk_score >= 8]),
            "event_types": self._group_events_by_type(last_24h),
            "timestamp": now.isoformat()
        }
        
    def _group_events_by_type(self, events: List[SecurityContext]) -> Dict[str, int]:
        """Group events by type"""
        counts = defaultdict(int)
        for event in events:
            counts[event.event_type.value] += 1
        return dict(counts)

class GrammarSecurityMiddleware:
    """Main security middleware for grammar service"""
    
    def __init__(self):
        self.store = InMemorySecurityStore()
        self.security_patterns = self._init_security_patterns()
        self.content_filters = self._init_content_filters()
        
    def _init_security_patterns(self) -> Dict[str, re.Pattern]:
        """Initialize security detection patterns"""
        return {
            "xss": re.compile(r'<script|javascript:|on\w+\s*=', re.IGNORECASE),
            "sql_injection": re.compile(r'\b(union|select|insert|delete|update|drop|create|alter|exec|execute)\b|(-{2})|(\|\|)', re.IGNORECASE),
            "path_traversal": re.compile(r'\.\.(\/|\\)'),
            "command_injection": re.compile(r'[;&|`$]', re.IGNORECASE)
        }
        
    def _init_content_filters(self) -> Dict[str, Any]:
        """Initialize content filtering rules"""
        return {
            "max_text_length": 10000,  # Maximum text length for grammar evaluation
            "min_text_length": 1,      # Minimum text length
            "max_words": 2000,         # Maximum number of words
            "suspicious_patterns": [
                r'(.)\1{20,}',         # Repeated characters (20+ times)
                r'\b\w{50,}\b',        # Very long words (50+ chars)
                r'[^\w\s\.,!?;:\'"-]{5,}',  # Multiple special characters
            ]
        }
        
    def get_client_ip(self, request: Request) -> str:
        """Extract client IP from request"""
        forwarded = request.headers.get("x-forwarded-for")
        if forwarded:
            return forwarded.split(",")[0].strip()
        real_ip = request.headers.get("x-real-ip")
        if real_ip:
            return real_ip
        return request.client.host if request.client else "unknown"
        
    def sanitize_text(self, text: str) -> str:
        """Sanitize input text"""
        if not isinstance(text, str):
            return str(text)
            
        # HTML escape
        text = html.escape(text)
        
        # Remove null bytes and control characters
        text = re.sub(r'[\x00-\x08\x0B-\x0C\x0E-\x1F\x7F]', '', text)
        
        # Normalize whitespace
        text = re.sub(r'\s+', ' ', text).strip()
        
        return text
        
    def detect_malicious_content(self, text: str) -> Dict[str, Any]:
        """Detect malicious content in text"""
        threats = []
        risk_score = 0
        
        for threat_type, pattern in self.security_patterns.items():
            if pattern.search(text):
                threats.append(threat_type)
                risk_score += 3
                
        # Check content filters
        if len(text) > self.content_filters["max_text_length"]:
            threats.append("oversized_content")
            risk_score += 2
            
        if len(text.split()) > self.content_filters["max_words"]:
            threats.append("too_many_words")
            risk_score += 2
            
        for pattern in self.content_filters["suspicious_patterns"]:
            if re.search(pattern, text):
                threats.append("suspicious_pattern")
                risk_score += 1
                
        return {
            "threats": threats,
            "risk_score": min(risk_score, 10),
            "safe": len(threats) == 0
        }
        
    def validate_language_code(self, lang: str) -> bool:
        """Validate language code"""
        # Common language codes for LanguageTool
        valid_langs = {
            'en', 'es', 'fr', 'de', 'it', 'pt', 'ru', 'zh', 'ja', 'ko',
            'en-US', 'en-GB', 'es-ES', 'fr-FR', 'de-DE', 'it-IT', 'pt-PT', 'pt-BR'
        }
        return lang in valid_langs
        
    def create_security_context(self, request: Request, event_type: GrammarSecurityEventType, 
                              risk_score: int, metadata: Dict[str, Any] = None) -> SecurityContext:
        """Create security context"""
        return SecurityContext(
            ip=self.get_client_ip(request),
            user_agent=request.headers.get("user-agent", ""),
            timestamp=datetime.now(),
            endpoint=str(request.url.path),
            method=request.method,
            risk_score=risk_score,
            event_type=event_type,
            metadata=metadata or {}
        )
        
    async def rate_limit_check(self, request: Request, 
                             max_requests: int = 30, window_seconds: int = 60) -> bool:
        """Check rate limiting"""
        ip = self.get_client_ip(request)
        
        # Check if IP is blocked
        if self.store.is_blocked(ip):
            context = self.create_security_context(
                request, GrammarSecurityEventType.RATE_LIMIT_EXCEEDED, 10,
                {"reason": "ip_blocked", "ip": ip}
            )
            self.store.log_security_event(context)
            return False
            
        # Check rate limit
        key = f"rate_limit:{ip}"
        result = self.store.track_request(key, window_seconds, max_requests)
        
        if not result["allowed"]:
            context = self.create_security_context(
                request, GrammarSecurityEventType.RATE_LIMIT_EXCEEDED, 6,
                {"requests": max_requests, "window": window_seconds}
            )
            self.store.log_security_event(context)
            return False
            
        return True
        
    async def validate_grammar_request(self, request: Request, body: Dict[str, Any]) -> Dict[str, Any]:
        """Validate grammar evaluation request"""
        errors = []
        risk_score = 0
        
        # Check required fields
        if "text" not in body:
            errors.append("Missing required field: text")
            risk_score += 2
            
        if "lang" not in body:
            errors.append("Missing required field: lang")
            risk_score += 1
            
        if errors:
            return {"valid": False, "errors": errors, "risk_score": risk_score}
            
        text = body.get("text", "")
        lang = body.get("lang", "")
        
        # Validate text
        if not isinstance(text, str):
            errors.append("Text must be a string")
            risk_score += 2
        elif len(text.strip()) == 0:
            errors.append("Text cannot be empty")
            risk_score += 1
        elif len(text) > self.content_filters["max_text_length"]:
            errors.append(f"Text too long (max {self.content_filters['max_text_length']} characters)")
            risk_score += 3
            
        # Validate language
        if not isinstance(lang, str):
            errors.append("Language must be a string")
            risk_score += 1
        elif not self.validate_language_code(lang):
            errors.append("Invalid language code")
            risk_score += 2
            
        # Check for malicious content
        if text:
            malicious_check = self.detect_malicious_content(text)
            if not malicious_check["safe"]:
                errors.extend([f"Threat detected: {threat}" for threat in malicious_check["threats"]])
                risk_score += malicious_check["risk_score"]
                
        return {
            "valid": len(errors) == 0,
            "errors": errors,
            "risk_score": risk_score
        }
        
    async def get_security_metrics(self) -> Dict[str, Any]:
        """Get security metrics"""
        metrics = self.store.get_metrics()
        metrics.update({
            "service": "grammar-service",
            "security_level": "high",
            "content_filtering": True,
            "rate_limiting": True,
            "threat_detection": True
        })
        return metrics

class GrammarSecurityHTTPMiddleware(BaseHTTPMiddleware):
    """HTTP middleware for grammar service security"""
    
    def __init__(self, app, security_middleware: GrammarSecurityMiddleware):
        super().__init__(app)
        self.security = security_middleware
        
    async def dispatch(self, request: Request, call_next):
        """Process request through security middleware"""
        start_time = time.time()
        
        # Apply security headers
        def add_security_headers(response: Response):
            response.headers.update({
                "X-Content-Type-Options": "nosniff",
                "X-Frame-Options": "DENY", 
                "X-XSS-Protection": "1; mode=block",
                "Strict-Transport-Security": "max-age=31536000; includeSubDomains",
                "Referrer-Policy": "strict-origin-when-cross-origin",
                "Content-Security-Policy": "default-src 'self'; script-src 'self'; style-src 'self'",
                "X-Service": "grammar-service",
                "X-Security-Level": "high"
            })
            return response
            
        # Rate limiting
        if not await self.security.rate_limit_check(request):
            response = JSONResponse(
                status_code=429,
                content={"error": "Rate limit exceeded", "retry_after": 60}
            )
            return add_security_headers(response)
            
        # Process request
        try:
            response = await call_next(request)
            
            # Log successful operation
            if request.url.path.startswith("/grammar/"):
                context = self.security.create_security_context(
                    request, GrammarSecurityEventType.GRAMMAR_EVALUATION, 1,
                    {
                        "status_code": response.status_code,
                        "response_time": time.time() - start_time,
                        "success": response.status_code < 400
                    }
                )
                self.security.store.log_security_event(context)
                
            return add_security_headers(response)
            
        except Exception as e:
            # Log error
            context = self.security.create_security_context(
                request, GrammarSecurityEventType.SUSPICIOUS_ACTIVITY, 7,
                {"error": str(e), "exception_type": type(e).__name__}
            )
            self.security.store.log_security_event(context)
            
            response = JSONResponse(
                status_code=500,
                content={"error": "Internal server error"}
            )
            return add_security_headers(response)

# Global security middleware instance
grammar_security = GrammarSecurityMiddleware()
